core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

```diff
@@ -219,8 +220,20 @@ private[spark] class SparkHadoopUtil extends Logging {
    */
   def listLeafStatuses(fs: FileSystem, baseStatus: FileStatus): Seq[FileStatus] = {
```

Comment:
if you want to list leaf nodes, then a listFiles(path, recursive=true) is way better. on s3 that's a single LIST request per 1000 files, irrespective of depth. and non "hallucinated" directories

Reply:
Cool. Let's try to use Hadoop's listFiles instead. In that case we don't have to do extra fsHasPathCapability check (in SparkHadoopUtil.scala at least)

Reply:
Alright. Could this be a compromise solution?
https://github.com/apache/spark/blob/afcfe22db8b208054f4887fa7c3bb4b90eee88f7/core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala#L221

Reply:
I can't see your previous change from the message. Could you point me to a particular Git commit?

Reply:
https://github.com/apache/spark/blob/afcfe22db8b208054f4887fa7c3bb4b90eee88f7/core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala#L221

Reply:
I actually reverted my commit, so it should be in this PR now